Add support for precision, scale, length to types as necessary.

robin900 opened the pull request:
Adds support for precision, scale, and other args/kwargs for reflected column type objects. Also, removes redundant column objects when view_columns contains all those
duplicates by using UNION instead of UNION ALL.

Fixes LocusEnergy/sqlalchemy-vertica-python#14.

robin900 committed:
Fix: remove redundant column objects when view_columns contains all those
duplicates by using UNION instead of UNION ALL. Add support for
precision, scale, etc. args and kwargs to reflected column type
objects.

Fixes LocusEnergy/sqlalchemy-vertica-python#14.

lemeryfertitta commented on sqla_vertica_python/vertica_python.py:
```diff
@@ -200,15 +200,54 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+                )
+            for row in connection.execute(column_select)
         ]
+        return [ c for c in colobjs if c ]
```
I think this is a non-optimal use of list comprehension.

```python
colobjs = []
for row in connection.execute(column_select):
  colobj = self._get_column_info(
    row.column_name, 
    row.data_type, 
    row.is_nullable,
    row.column_default,
    (row.column_name in primary_key_columns)
  )
  if colobj:
    colobjs.append(colobj)
```

Skips a full list build and iteration

lemeryfertitta commented on sqla_vertica_python/vertica_python.py:
```diff
@@ -200,15 +200,54 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+            typekwargs['timezone'] = True
+
+        if callable(typeobj):
+            typeargs = tuple(typeargs)
```
This line is unnecessary. `*` will unpack a list the same as a tuple

lemeryfertitta commented on sqla_vertica_python/vertica_python.py:
```diff
@@ -200,15 +200,54 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+
+    def _get_column_info(self, name, data_type, is_nullable, default, is_primary_key):
+        m = re.match(r'(\w+)(?:\((\d+)(?:,(\d+))?\))?', data_type)
+        if not m:
```
What isn't being matched? A quick glance makes me think this should probably raise an error, so this check may not be needed.

lemeryfertitta commented on sqla_vertica_python/vertica_python.py:
```diff
@@ -200,15 +200,54 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+            except ValueError:
+                pass
+
+        if 'TIMEZONE' in typename or 'TIME ZONE' in typename:
```
We should add `TIMESTAMPTZ` to this, and refactor to something like the following:

```python
if any(tz_string in typename for tz_string in ('TIMEZONE', 'TIME ZONE', 'TIMESTAMPTZ'))
```

lemeryfertitta commented on sqla_vertica_python/vertica_python.py:
```diff
@@ -200,15 +200,54 @@ def get_columns(self, connection, table_name, schema=None, **kw):
+                v = int(m.group(3))
+                typeargs.append(v)
+            except ValueError:
+                pass
```
Not a huge fan of this section, its longer and more complicated than necessary. Here is a suggeseted refactor that should be functionally equivalent:

```python
for arg_group in (2, 3):
  try:
    typeargs.append(int(m.group(arg_group)))

  except (TypeError, ValueError):
   pass
```

lemeryfertitta reviewed (changes requested):
Good improvement, thanks for the PR. Once the review comments are addressed, I'll package this up into a new release.

robin900 committed:
made requested changes, with one adjustment: check m.group() value
for truth before attempting the int(), since unwinding 1-2 exceptions
per column slows down the reflection process, visibly for wide tables.

lemeryfertitta commented:
Thanks for the updates @robin900 - I'll get this released momentarily
